File: orders/src/utils/telemetry.py

import os

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import (
    BatchSpanProcessor,
    ConsoleSpanExporter,
)
from opentelemetry.sdk.resources import Resource

from .logger import Logger
from .singleton import Singleton

# otlp => open telemetry specification


SERVICE_NAME = "order_ms"


class Telemetry(metaclass=Singleton):
    """"
        Telemetry 
    """

    def __init__(self) -> None:
        """Initialize"""
        self.service_instance = os.getenv("HOSTNAME")
        self.provider = None
        self.logger = Logger().get_logger()

    def setup_telemetry_tracer_provider(self):
        """
            configure opentelemetry sdk
        """
        self.provider = TracerProvider(
            resource=Resource.create({
                "service.namespace": "shop",
                "service.name": SERVICE_NAME,
                "service.instance.id": self.service_instance,
                "service.version": "v1.0.0",
            }),
        )
        processor = BatchSpanProcessor(ConsoleSpanExporter())
        self.provider.add_span_processor(processor)

        # Sets the global default tracer provider
        trace.set_tracer_provider(self.provider)

    def close_telemetry_tracer_provider(self):
        """
            shutdown tracer provider
        """
        self.provider.shutdown()

File: consumers/src/lib/telemetry.cpp

#include <opentelemetry/exporters/otlp/otlp_grpc_exporter.h>
#include <opentelemetry/sdk/trace/samplers/always_on.h>
#include <opentelemetry/sdk/trace/processor.h>
#include <opentelemetry/sdk/trace/simple_processor.h>
#include <opentelemetry/sdk/trace/tracer_provider.h>
#include <opentelemetry/trace/provider.h>
#include <opentelemetry/sdk/trace/batch_span_processor.h>

void setUpTelemetry()
{
    // resources
    auto resource_attributes = opentelemetry::sdk::resource::ResourceAttributes{
        {"service.namespace", "shop"},
        {"service.name", std::getenv("SERVICE_NAME")},
        {"service.instance.id", std::getenv("HOSTNAME")},
        {"service.version", "v1.0.0"}};
    auto resource = opentelemetry::sdk::resource::Resource::Create(resource_attributes);

    // exporter
    opentelemetry::exporter::otlp::OtlpGrpcExporterOptions opts;
    opts.endpoint = std::getenv("LOG_COLLECTOR_ENDPOINT");
    opts.use_ssl_credentials = true;
    opts.ssl_credentials_cacert_as_string = "ssl-certificate";
    auto exporter =
        std::unique_ptr<opentelemetry::sdk::trace::SpanExporter>(opentelemetry::exporter::otlp::OtlpGrpcExporter(opts));

    // exporter
    opentelemetry::sdk::trace::BatchSpanProcessorOptions options;
    auto processor = std::unique_ptr<opentelemetry::sdk::trace::SpanProcessor>(
        opentelemetry::sdk::trace::BatchSpanProcessor(std::move(exporter), options));

    // sampler
    // auto sampler = std::unique_ptr<opentelemetry::sdk::trace::Sampler>(new opentelemetry::sdk::trace::AlwaysOnSampler);

    // trace provider
    auto provider = opentelemetry::nostd::shared_ptr<opentelemetry::trace::TracerProvider>(opentelemetry::sdk::trace::TracerProvider(std::move(processor),
                                                                                                                                     resource));

    // set the global tracer TraceProvider
    opentelemetry::trace::Provider::SetTracerProvider(std::move(provider));
}